README.md
# tilt-wizard

[![Build
Status](https://dev.azure.com/johnstrunk/tilt-wizard/_apis/build/status/JohnStrunk.tilt-wizard)](https://dev.azure.com/johnstrunk/tilt-wizard/_build/latest?definitionId=1)

`tilt-wizard` is a simple Windows cli program to automatically calibrate
DirectInput-based motion sensors used for Virtual Pinball.

## Background

When building a virtual pinball cabinet, a popular option to add is a motion
sensor so that it is possible to "nudge" the table while playing and have this
real-life jostling of the table affect the movement of the virtual pinball.

When I built my table, I chose to use the [Mot-Ion
controller](http://www.nanotechgaming.com/mot-ionkit.php). It appears to the
computer as a gamepad with the x and y axes detecting motion of the cabinet.
Unfortunately, these units have a good bit of noise to them, are difficult to
calibrate, and their center point tends to change with temperature.

This project is my attempt to create a simple executable that automatically
calibrates the center, limits, and deadzone on-the-fly.

## Using tilt-wizard

Download the latest version from the [Releases
page](https://github.com/JohnStrunk/tilt-wizard/releases).

Run `tilt-wizard -l` with no arguments to list available DirectInput devices:

```
C:\tilt-wizard>tilt-wizard.exe -l
```

This should discover any devices attached to your computer. To have tilt-wizard
automatically tune the device, pass in the GUID of the desired device via the
`-d` option:

```
C:\tilt-wizard>tilt-wizard.exe -d {34BE2320-42C7-11E6-8001-444553540000}
```

The application will then run, sampling the X & Y axes and updating the center
point.

- `current`: The most recent sample
- `avg`: The currently calculated average (and center point)
- `DZ%`: The recommended deadzone percentage for this axis. It is calculated
  based on the standard deviation of the incoming data.

Example:

![Example output for tilt-wizard](docs/images/output.png)

To integrate tilt-wizard with your pinball cabinet, see the [installation
document](docs/installation.md).

## Building from source

You will need [MinGW-w64](http://www.mingw-w64.org/) to build the source. While
you can download and install it directly, I recommend using
[Chocolatey](https://chocolatey.org/). Chocolatey is [easy to
install](https://chocolatey.org/install#install-with-cmdexe) and once you get it
running, you can install MinGW-w64 via:

```
C:\>choco install mingw
...
```

Once you have the prerequisites installed, make sure you update your PATH. In my
case, that was:

```
set PATH=%PATH%;C:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64\bin
```

You should now be able to build via:

```
C:\tilt-wizard>mingw32-make
... build output ...
```

Then run it via:

```
C:\tilt-wizard>tilt-wizard.exe
```


CHANGELOG.md
# Changelog

## Master

## Version v1.0

- Added
  - Axis range can be customized via `-r range`.
  - Moving average momentum can be customized via `-m value`.
  - `direader.exe` to help diagnose add debug DirectInput settings
- Changed
  - Command line has changed to use switches. (`-l` to list devices and `-d
    guid` to tune a particular device).
  - Output now shows recommended deadzone for each axis instead of the standard
    deviation of incoming data.
- Limitations
  - Deadzone settings are not visible to other applications and must be set
    manually.
  - Changes to the calibration are not visible to applications that are
    currently running. Only newly launched applications see the most current
    configuration.

## Version v0.1

- Added
  - Initial release.
  - Ability to list available devices.
  - Ability to automatically tune center point of X & Y axes.


Device.h
/*
 * Tilt-wizard
 * Copyright (C) 2018  John D. Strunk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

 #ifndef DEVICE_H
 #define DEVICE_H

#include <list>
#include <string>

#define DIRECTINPUT_VERSION 0x0800
#include <dinput.h>
#include "dinputd.h"

class Device {
public:
    enum CalibrationMode { RAW, COOKED };

private:
    /// Pointer to the underlying DirectInput device
    LPDIRECTINPUTDEVICE8 _dev;
    LPDIJOYCONFIG _jc;
    LPDIRECTINPUTJOYCONFIG _dijc;
    LPDIRECTINPUTJOYCONFIG8 _dijc8;
    /// Most recently polled state
    DIJOYSTATE _state;

    void _init(IID *guid, CalibrationMode mode);

public:
    /**
     * Create a new Device object from its GUID
     */
    Device(IID *guid, CalibrationMode mode = RAW);
    /// Create a new Device object from its GUID passed as a string
    Device(std::string guidString, CalibrationMode mode = RAW);
    ~Device();

    struct DeviceDescription {
        std::string deviceName;
        std::string guidString;
        IID guid;
    };

    typedef std::list<DeviceDescription> DescriptionList;

    static DescriptionList enumerateDevices();

    /**
     * Get the instance name for the device
     */
    std::string name() const;

    /**
     * Poll the latest state (position) of the device. This will cause all
     * associated Axis objects to get updated values.
     */
    void poll();

    void calibration(WORD axisOffset, LONG *min, LONG *center, LONG *max) const;
    void calibration(WORD axisOffset, LONG min, LONG center, LONG max);
    double deadzone(WORD axisOffset) const;
    void deadzone(WORD axisOffset, double pct);
    LONG position(WORD axisOffset) const;
    void range(WORD axisOffset, LONG min, LONG max);
    void range(WORD axisOffset, LONG *min, LONG *max) const;
    double saturation(WORD axisOffset) const;
    void saturation(WORD axisOffset, double pct);
};

 #endif // DEVICE_H


EMStat.h
/*
 * Tilt-wizard
 * Copyright (C) 2018  John D. Strunk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef EMSTAT_H
#define EMSTAT_H

#include <math.h>

/// Calculates exponential moving statistics
class EMStat {
    /// Weight of the existing statistic (0-1.0)
    double _momentum;
    /// Current average
    double _avg;
    /// Current variance
    double _var;
public:
    /// Create a new stat object using the specified momentum
    EMStat(double momentum): _momentum(momentum), _avg(0), _var(0) { }

    /// Reset the statistics to a specified value
    void set(double avg, double var) { _avg = avg; _var = var; }
    /// Insert a new data point
    void insert(double val);

    double avg() const { return _avg; }
    double var() const { return _var; }
    double stdev() const { return sqrt(var()); }
};

#endif // EMSTAT_H


Device.cc
/*
 * Tilt-wizard
 * Copyright (C) 2018  John D. Strunk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// Docs: https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ee417929(v%3dvs.85)

#include "Device.h"
#include "TWError.h"

BOOL
devEnumCb(LPCDIDEVICEINSTANCE lpddi, LPVOID voidDev)
{
    Device::DescriptionList *devices =
        reinterpret_cast<Device::DescriptionList*>(voidDev);
    const unsigned G_LEN = 100;
    wchar_t guid[G_LEN];
    if (0 == StringFromGUID2(lpddi->guidInstance, guid, G_LEN))
        throw TWError("Unable to convert GUID to string");
    std::wstring wgstring(guid);
    Device::DeviceDescription desc;
    desc.deviceName = lpddi->tszInstanceName;
    desc.guidString.assign(wgstring.begin(), wgstring.end());
    memcpy(&desc.guid, &lpddi->guidInstance, sizeof(desc.guid));
    devices->push_back(desc);
    return DIENUM_CONTINUE;
}

Device::DescriptionList
Device::enumerateDevices()
{
    // Initialize module handle instance
    HINSTANCE hInst = GetModuleHandle(0);
    if (!hInst) throw TWError("Failed to get module handle");

    // Initialize DirectInput
    LPDIRECTINPUT8 di8Interface = 0;
    HRESULT res = DirectInput8Create(hInst, DIRECTINPUT_VERSION,
                                     IID_IDirectInput8, (LPVOID*)&di8Interface,
                                     0);
    if (FAILED(res)) throw TWError("Failed to get interface to DirectInput", res);

    DescriptionList devices;
    res = di8Interface->EnumDevices(DI8DEVCLASS_GAMECTRL, devEnumCb,
                                    &devices, DIEDFL_ATTACHEDONLY);
    if (FAILED(res)) throw TWError("Error enumerating devices", res);

    return devices;
}


Device::Device(std::string guidString, CalibrationMode mode)
{
    IID guid;
    std::wstring wGuid;
    wGuid.assign(guidString.begin(), guidString.end());
    HRESULT res = IIDFromString(wGuid.c_str(), &guid);
    if (FAILED(res)) throw TWError("Error parsing device GUID", res);

    _init(&guid, mode);
}

Device::Device(IID *guid, CalibrationMode mode)
{
    _init(guid, mode);
}

void
Device::_init(IID *guid, CalibrationMode mode)
{
    if (!guid) throw TWError("GUID should not be null");

    HINSTANCE hInst = GetModuleHandle(0);
    if (!hInst) throw TWError("Unable to get module handle");

    LPDIRECTINPUT8 di8Int = 0;
    HRESULT res = DirectInput8Create(hInst, DIRECTINPUT_VERSION,
                                     IID_IDirectInput8, (LPVOID*)&di8Int,
                                     0);
    if (FAILED(res)) throw TWError("Failed to get interface to DirectInput", res);

    res = di8Int->QueryInterface(IID_IDirectInputJoyConfig8, (LPVOID*)&_dijc8);
    if (FAILED(res)) throw TWError("Failed to get JoyConfig8", res);
    res = _dijc8->SetCooperativeLevel(GetConsoleWindow(), DISCL_BACKGROUND | DISCL_EXCLUSIVE);
    if (FAILED(res)) throw TWError("SetCooperativeLevel", res);
    res = _dijc8->Acquire();
    if (FAILED(res)) throw TWError("Acquire", res);
    res = _dijc8->SendNotify();
    if (FAILED(res)) throw TWError("SendNotify", res);

    res = di8Int->CreateDevice(*guid, &_dev, 0);
    if (FAILED(res)) throw TWError("Unable to get handle for device", res);

    di8Int->Release();

    // Configure the device
    res = _dev->SetCooperativeLevel(0, DISCL_BACKGROUND | DISCL_NONEXCLUSIVE);
    if (FAILED(res)) throw TWError("Failed setting device cooperation level", res);
    res = _dev->SetDataFormat(&c_dfDIJoystick);
    if (FAILED(res)) throw TWError("Failed setting data format", res);
    // Set to retrieve RAW data from device
    DIPROPDWORD dipdw;
    dipdw.diph.dwSize = sizeof(dipdw);
    dipdw.diph.dwHeaderSize = sizeof(dipdw.diph);
    dipdw.diph.dwObj = 0;
    dipdw.diph.dwHow = DIPH_DEVICE;
    if (mode == RAW) {
        dipdw.dwData = DIPROPCALIBRATIONMODE_RAW;
    } else {
        dipdw.dwData = DIPROPCALIBRATIONMODE_COOKED;
    }
    res = _dev->SetProperty(DIPROP_CALIBRATIONMODE, &dipdw.diph);
    if (FAILED(res)) throw TWError("Failed setting calibration mode", res);

    // Open the device so we can read
    res = _dev->Acquire();
    if (FAILED(res)) throw TWError("Failed acquiring device", res);
}

Device::~Device()
{
    _dev->Unacquire();
    _dev->Release(); // Needed?
}

std::string
Device::name() const
{
    DIDEVICEINSTANCE di;
    di.dwSize = sizeof(di);
    HRESULT res = _dev->GetDeviceInfo(&di);
    if (FAILED(res)) throw TWError("Failed getting device info", res);
    std::string name(di.tszInstanceName);
    return name;
}

void
Device::poll()
{
    HRESULT res = _dev->Poll();
    if (FAILED(res)) throw TWError("Failed polling device", res);
    res = _dev->GetDeviceState(sizeof(_state), &_state);
    if (FAILED(res)) throw TWError("Failed getting current device position", res);
}

void
Device::calibration(WORD axisOffset, LONG *min, LONG *center, LONG *max) const
{
    DIPROPCAL dical;
    dical.diph.dwSize = sizeof(dical);
    dical.diph.dwHeaderSize = sizeof(dical.diph);
    dical.diph.dwObj = axisOffset;
    dical.diph.dwHow = DIPH_BYOFFSET;
    HRESULT res = _dev->GetProperty(DIPROP_CALIBRATION, &dical.diph);
    if (FAILED(res)) throw TWError("Unable to get calibration", res);
    *min = dical.lMin;
    *center = dical.lCenter;
    *max = dical.lMax;
}

void
Device::calibration(WORD axisOffset, LONG min, LONG center, LONG max)
{
    DIPROPCAL dical;
    dical.diph.dwSize = sizeof(dical);
    dical.diph.dwHeaderSize = sizeof(dical.diph);
    dical.diph.dwObj = axisOffset;
    dical.diph.dwHow = DIPH_BYOFFSET;
    dical.lMin = min;
    dical.lCenter = center;
    dical.lMax = max;
    HRESULT res = _dev->SetProperty(DIPROP_CALIBRATION, &dical.diph);
    if (FAILED(res)) throw TWError("Unable to set calibration", res);

/*
    LPDIRECTINPUTJOYCONFIG idijc = 0;
    res = di8Dev->QueryInterface(IID_IDirectInputJoyConfig,(LPVOID*)&idijc);
    if (FAILED(res)) fatalError(L"Unable to get JoyConfig", res);
    res = idijc->Acquire();
    if (FAILED(res)) fatalError(L"JoyConfig acquire", res);
    res = idijc->SendNotify();
    if (FAILED(res)) fatalError(L"JoyConfig SendNotify", res);
    res = idijc->Release();
    if (FAILED(res)) fatalError(L"JoyConfig Release", res);
*/
}

LONG
Device::position(WORD axisOffset) const
{
    switch (axisOffset) {
    case DIJOFS_X:
        return _state.lX;
    case DIJOFS_Y:
        return _state.lY;
    default:
        throw TWError("Unsupported axis");
    }

    // not reached
    return 0;
}

double
Device::deadzone(WORD axisOffset) const
{
    DIPROPDWORD dipdw;
    dipdw.diph.dwSize = sizeof(dipdw);
    dipdw.diph.dwHeaderSize = sizeof(dipdw.diph);
    dipdw.diph.dwObj = axisOffset;
    dipdw.diph.dwHow = DIPH_BYOFFSET;
    HRESULT res = _dev->GetProperty(DIPROP_DEADZONE, &dipdw.diph);
    if (FAILED(res)) throw TWError("Unable to get deadzone", res);
    return dipdw.dwData/100.0;
}

void
Device::deadzone(WORD axisOffset, double pct)
{
    if (0.0 > pct || pct > 100.0)
        throw TWError("Deadzone must be between 0 and 100");
    DIPROPDWORD dipdw;
    dipdw.diph.dwSize = sizeof(dipdw);
    dipdw.diph.dwHeaderSize = sizeof(dipdw.diph);
    dipdw.diph.dwObj = axisOffset;
    dipdw.diph.dwHow = DIPH_BYOFFSET;
    dipdw.dwData = (DWORD)(pct*100.0);
    HRESULT res = _dev->SetProperty(DIPROP_DEADZONE, &dipdw.diph);
    if (FAILED(res)) throw TWError("Unable to set deadzone", res);
}

void
Device::range(WORD axisOffset, LONG min, LONG max)
{
    if (min > max) throw TWError("min must be <= max");
    DIPROPRANGE r;
    r.diph.dwSize = sizeof(r);
    r.diph.dwHeaderSize = sizeof(r.diph);
    r.diph.dwObj = axisOffset;
    r.diph.dwHow = DIPH_BYOFFSET;
    r.lMin = min;
    r.lMax = max;
    HRESULT res = _dev->SetProperty(DIPROP_RANGE, &r.diph);
    if (FAILED(res)) throw TWError("Unable to set axis range", res);
}

void
Device::range(WORD axisOffset, LONG *min, LONG *max) const
{
    DIPROPRANGE r;
    r.diph.dwSize = sizeof(r);
    r.diph.dwHeaderSize = sizeof(r.diph);
    r.diph.dwObj = axisOffset;
    r.diph.dwHow = DIPH_BYOFFSET;
    HRESULT res = _dev->GetProperty(DIPROP_RANGE, &r.diph);
    if (FAILED(res)) throw TWError("Unable to get axis range", res);
    *min = r.lMin;
    *max = r.lMax;
}

double
Device::saturation(WORD axisOffset) const
{
    DIPROPDWORD dipdw;
    dipdw.diph.dwSize = sizeof(dipdw);
    dipdw.diph.dwHeaderSize = sizeof(dipdw.diph);
    dipdw.diph.dwObj = axisOffset;
    dipdw.diph.dwHow = DIPH_BYOFFSET;
    HRESULT res = _dev->GetProperty(DIPROP_SATURATION, &dipdw.diph);
    if (FAILED(res)) throw TWError("Unable to get axis saturation", res);
    return dipdw.dwData/100.0;
}

void
Device::saturation(WORD axisOffset, double pct)
{
    if (0.0 > pct || pct > 100.0)
        throw TWError("Saturation must be between 0 and 100");
    DIPROPDWORD dipdw;
    dipdw.diph.dwSize = sizeof(dipdw);
    dipdw.diph.dwHeaderSize = sizeof(dipdw.diph);
    dipdw.diph.dwObj = axisOffset;
    dipdw.diph.dwHow = DIPH_BYOFFSET;
    dipdw.dwData = (DWORD)(pct*100.0);
    HRESULT res = _dev->SetProperty(DIPROP_SATURATION, &dipdw.diph);
    if (FAILED(res)) throw TWError("Unable to set axis saturation", res);
}


EMStat.cc
/*
 * Tilt-wizard
 * Copyright (C) 2018  John D. Strunk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "EMStat.h"

void
EMStat::insert(double val)
{
        double pavg = _avg;
        _avg = _momentum * pavg + (1 - _momentum) * val;

        double pvar = _var;
        _var = _momentum * pvar +
               (1 - _momentum) * (val - pavg) * (val - _avg);
}


TWError.h
/*
 * Tilt-wizard
 * Copyright (C) 2018  John D. Strunk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

 #ifndef TWERROR_H
 #define TWERROR_H

#include <comdef.h>
#include <stdexcept>

class TWError : public std::runtime_error {
public:
    HRESULT error;

    TWError(const std::string &msg, HRESULT err) :
        std::runtime_error(msg + ": " + _com_error(err).ErrorMessage()),
        error(err) { }
    TWError(const std::string &msg) :
        std::runtime_error(msg), error(E_FAIL) { }
};

 #endif // TWERROR_H


direader.cc
/*
 * Tilt-wizard
 * Copyright (C) 2018  John D. Strunk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <iostream>
#include <string>
#include <getopt.h>
#include <iomanip>
#include <unistd.h>

#include "Device.h"

#define _TEXTIFY1(x) #x
#define TEXTIFY(x) _TEXTIFY1(x)

static void
enumerateDevices()
{
    Device::DescriptionList devs = Device::enumerateDevices();

    std::cout << std::setw(42) << std::left << "Device GUID"
               << "Name" << std::endl;
    std::cout << std::setw(70) << std::setfill('=') << "=" << std::endl;

    for (Device::DescriptionList::const_iterator i = devs.begin();
         i != devs.end(); ++i) {
        std::cout << i->guidString << "    " << i->deviceName << std::endl;
    }
}

void
readDevice(std::string guidString)
{
    std::cout << "Looking for device: " << guidString << std::endl;
    Device dev(guidString);
    Device cooked(guidString, Device::COOKED);
    std::cout << "Device name: " << dev.name() << std::endl;

    unsigned fwidth = 6;
    std::cout << std::right
              << std::setw(2*fwidth) << "cooked"
              << std::setw(fwidth) << "raw"
              << std::setw(fwidth) << "rmin"
              << std::setw(fwidth) << "min"
              << std::setw(fwidth) << "ctr"
              << std::setw(fwidth) << "max"
              << std::setw(fwidth) << "rmax"
              << std::setw(fwidth) << "DZ%"
              << std::setw(fwidth) << "SAT%"
              << std::setw(2*fwidth) << "cooked"
              << std::setw(fwidth) << "raw"
              << std::setw(fwidth) << "rmin"
              << std::setw(fwidth) << "min"
              << std::setw(fwidth) << "ctr"
              << std::setw(fwidth) << "max"
              << std::setw(fwidth) << "rmax"
              << std::setw(fwidth) << "DZ%"
              << std::setw(fwidth) << "SAT%"
              << std::endl;

    while (true) {
        dev.poll();
        cooked.poll();

        LONG xMin, xCtr, xMax;
        dev.calibration(DIJOFS_X, &xMin, &xCtr, &xMax);
        LONG xrMin, xrMax;
        dev.range(DIJOFS_X, &xrMin, &xrMax);

        LONG yMin, yCtr, yMax;
        dev.calibration(DIJOFS_Y, &yMin, &yCtr, &yMax);
        LONG yrMin, yrMax;
        dev.range(DIJOFS_Y, &yrMin, &yrMax);

        std::cout << std::setprecision(0) << std::fixed;
        std::cout << std::setw(fwidth) << "x: "
                  << std::setw(fwidth) << cooked.position(DIJOFS_X)
                  << std::setw(fwidth) << dev.position(DIJOFS_X)
                  << std::setw(fwidth) << xrMin
                  << std::setw(fwidth) << xMin
                  << std::setw(fwidth) << xCtr
                  << std::setw(fwidth) << xMax
                  << std::setw(fwidth) << xrMax
                  << std::setw(fwidth) << dev.deadzone(DIJOFS_X)
                  << std::setw(fwidth) << dev.saturation(DIJOFS_X)
                  << std::setw(fwidth) << "y: "
                  << std::setw(fwidth) << cooked.position(DIJOFS_Y)
                  << std::setw(fwidth) << dev.position(DIJOFS_Y)
                  << std::setw(fwidth) << yrMin
                  << std::setw(fwidth) << yMin
                  << std::setw(fwidth) << yCtr
                  << std::setw(fwidth) << yMax
                  << std::setw(fwidth) << yrMax
                  << std::setw(fwidth) << dev.deadzone(DIJOFS_Y)
                  << std::setw(fwidth) << dev.saturation(DIJOFS_Y)
                  << "\r";
        usleep(100000);
    }
}


static void
usage(std::string pname)
{
    std::cout << std::endl << "Usage:" << std::endl
              << "  " << pname.c_str() << " -h|--help" << std::endl
              << "  " << pname.c_str() << " -l|--list" << std::endl
              << "  " << pname.c_str() << " -d|--device device_guid" << std::endl
              << std::endl
              << "  -d, --device device_uuid   read from specified device"
              << std::endl
              << "  -h, --help                 this help message"
              << std::endl
              << "  -l, --list                 list available devices"
              << std::endl;
}


 static void
 printVersion()
 {
     std::cout << "direader - https://github.com/JohnStrunk/tilt-wizard"
               << std::endl
               << "License: AGPL v3 or later" << std::endl
               << "Version: " << TEXTIFY(GIT_VERSION) << std::endl
               << "Date: " << TEXTIFY(GIT_DATE) << std::endl << std::endl;
 }


 int main(int argc, char *argv[])
 {
     printVersion();

     std::string devGuid;
     char opt;
     struct option longopts[] = {
         {"device", required_argument, 0, 'd'},
         {"list", no_argument, 0, 'l'},
         {"help", no_argument, 0, 'h'},
         {0, 0, 0, 0}
     };
     while ((opt = getopt_long(argc, argv, "d:hl", longopts, 0)) != -1) {
         switch (opt) {
         case 'd':
             devGuid = optarg;
             break;
         case 'h':
             usage(argv[0]);
             exit(EXIT_SUCCESS);
             break;
         case 'l':
             enumerateDevices();
             exit(EXIT_SUCCESS);
             break;
         default: /* '?' */
             usage(argv[0]);
             exit(EXIT_FAILURE);
         }
     }

     if (!devGuid.empty()) {
         readDevice(devGuid);
         return 0;
     } else {
         std::cout << "Must specify one of -d or -h" << std::endl;
         usage(argv[0]);
         exit(EXIT_FAILURE);
     }
 }


tilt-wizard.cc
/*
 * Tilt-wizard
 * Copyright (C) 2018  John D. Strunk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <algorithm>
#include <cstdio>
#include <iomanip>
#include <iostream>
#include <ostream>
#include <string>
#include <getopt.h>
#include <unistd.h>

#define DIRECTINPUT_VERSION 0x0800
#include <comdef.h>
#include <dinput.h>
//#include <dinputd.h>
#include <Windows.h>

#include "Device.h"
#include "EMStat.h"
#include "TWError.h"

#ifndef GIT_VERSION
#define GIT_VERSION (unknown)
#endif

#ifndef GIT_DATE
#define GIT_DATE (unknown)
#endif

#define _TEXTIFY1(x) #x
#define TEXTIFY(x) _TEXTIFY1(x)

static const int DEFAULT_RANGE = 100;
static const double DEFAULT_MOMENTUM = 0.95;

static void
enumerateDevices()
{
    Device::DescriptionList devs = Device::enumerateDevices();

    std::cout << std::setw(42) << std::left << "Device GUID"
               << "Name" << std::endl;
    std::cout << std::setw(70) << std::setfill('=') << "=" << std::endl;

    for (Device::DescriptionList::const_iterator i = devs.begin();
         i != devs.end(); ++i) {
        std::cout << i->guidString << "    " << i->deviceName << std::endl;
    }
}

static void
calibrateDevice(std::string guidString, int axisRange, double momentum)
{
    std::cout << "Looking for device: " << guidString << std::endl;
    Device dev(guidString);
    std::cout << "Device name: " << dev.name() << std::endl;

    // Grab values once to initialize the stats
    dev.poll();
    sleep(1);  // Let the device start sending data

    /*
     * Create stats objects
     * We take 10 samples/sec, and the new data has weight 1%, so each
     * second is ~10% of the statistic
     */
    EMStat xStats(momentum), yStats(momentum);
    dev.poll();
    xStats.set(dev.position(DIJOFS_X), 100*100);
    yStats.set(dev.position(DIJOFS_Y), 100*100);

    std::cout << std::right
              << std::setw(13) << "current"
              << std::setw(7) << "avg"
              << std::setw(7) << "DZ%"
              << std::setw(13) << "current"
              << std::setw(7) << "avg"
              << std::setw(7) << "DZ%"
              << std::endl;

    for (unsigned i=1; ;++i) {
        dev.poll();

        LONG x = dev.position(DIJOFS_X);
        // Reject extreme values (nudges); ensure we always accept something
        if (abs(x - xStats.avg()) < std::max(2.0*xStats.stdev(), 10.0))
            xStats.insert(x);

        LONG y = dev.position(DIJOFS_Y);
        if (abs(y - yStats.avg()) < std::max(2.0*yStats.stdev(), 10.0))
            yStats.insert(y);

        double xDeadzone = std::min(4.0*xStats.stdev()*100.0/axisRange, 100.0);
        double yDeadzone = std::min(4.0*yStats.stdev()*100.0/axisRange, 100.0);

        // Update calibration 1/sec
        if (i % 10 == 0) {
            dev.deadzone(DIJOFS_X, xDeadzone);
            dev.calibration(DIJOFS_X, xStats.avg() - axisRange,
                                      xStats.avg(),
                                      xStats.avg() + axisRange);
            dev.deadzone(DIJOFS_Y, yDeadzone);
            dev.calibration(DIJOFS_Y, yStats.avg() - axisRange,
                                      yStats.avg(),
                                      yStats.avg() + axisRange);
        }

        // Print stats
        std::cout << std::setprecision(0) << std::fixed;
        std::cout << "\r    x: " << std::setw(6) << x
                  << " " << std::setw(6) << xStats.avg()
                  << " " << std::setw(6) << xDeadzone
                  << "    y: " << std::setw(6) << y
                  << " " << std::setw(6) << yStats.avg()
                  << " " << std::setw(6) << yDeadzone;
        // Pause 0.1s between samples
        usleep(100000);
    }
}


static void
usage(std::string pname)
{
    std::cout << std::endl << "Usage:" << std::endl
              << "  " << pname.c_str() << " -h|--help" << std::endl
              << "  " << pname.c_str() << " -l|--list" << std::endl
              << "  " << pname.c_str() << " -d|--device device_guid  [options...]" << std::endl
              << std::endl
              << "  -d, --device device_uuid   auto-calibrate specified device"
              << std::endl
              << "  -h, --help                 this help message"
              << std::endl
              << "  -l, --list                 list available devices"
              << std::endl
              << "  -m, --momentum             EMA momentum (0 - 1.0)"
              << " (default: " << DEFAULT_MOMENTUM << ")"
              << std::endl
              << "  -r, --range                axis range, center to max"
              << " (default: " << DEFAULT_RANGE << ")"
              << std::endl;
}


static void
printVersion()
{
    std::cout << "tilt-wizard - https://github.com/JohnStrunk/tilt-wizard"
              << std::endl
              << "License: AGPL v3 or later" << std::endl
              << "Version: " << TEXTIFY(GIT_VERSION) << std::endl
              << "Date: " << TEXTIFY(GIT_DATE) << std::endl << std::endl;
}


int main(int argc, char *argv[])
{
    printVersion();

    std::string devGuid;
    int range = DEFAULT_RANGE;
    double momentum = DEFAULT_MOMENTUM;
    char opt;
    struct option longopts[] = {
        {"device", required_argument, 0, 'd'},
        {"help", no_argument, 0, 'h'},
        {"list", no_argument, 0, 'l'},
        {"momentum", required_argument, 0, 'm'},
        {"range", required_argument, 0, 'r'},
        {0, 0, 0, 0}
    };
    while ((opt = getopt_long(argc, argv, "d:hlm:r:", longopts, 0)) != -1) {
        switch (opt) {
        case 'd':
            devGuid = optarg;
            break;
        case 'h':
            usage(argv[0]);
            exit(EXIT_SUCCESS);
            break;
        case 'l':
            enumerateDevices();
            exit(EXIT_SUCCESS);
            break;
        case 'm':
            momentum = atof(optarg);
            break;
        case 'r':
            range = atoi(optarg);
            break;
        default: /* '?' */
            usage(argv[0]);
            exit(EXIT_FAILURE);
        }
    }

    if (momentum < 0 || 1.0 < momentum) {
        std::cout << "Momentum must be in the range 0 - 1.0" << std::endl;
        usage(argv[0]);
        exit(EXIT_FAILURE);
    }

    if (range <= 0) {
        std::cout << "Axis range must be > 0" << std::endl;
        usage(argv[0]);
        exit(EXIT_FAILURE);
    }

    if (!devGuid.empty()) {
        calibrateDevice(devGuid, range, momentum);
        return 0;
    } else {
        std::cout << "Must specify one of -d, -h or -l" << std::endl;
        usage(argv[0]);
        exit(EXIT_FAILURE);
    }
}
